DRAFT: Adds web command

MatheusRich opened the pull request:
# Description:

Closes #2784.

Ruby claims to be "a programmer's best friend", and focus on "simplicity and productivity". As rubygems is now part of Ruby, I think it should follow those principles too. 

I'm integrating [gem-web](https://github.com/bitboxer/gem-web) into rubygems as part of [GSoC 2019](https://summerofcode.withgoogle.com/projects/#5260793296191488). This new command allows user to quickly find documentation of a gem.

I think this is specially good when one's not sure about what a gem does, or is confused with its name. Maybe the user wants to quickly check the rubygems' page of some gem (to check versions, dependencies, etc). With this command the access will be faster and easier with just one command on the terminal. Other use case could be opening the documentation for a gem, which improves productivity.

The usage would be like this:
```
gem web rails # Opens github
```

    Options:
      -g, --github                     Open GitHub page of gem, this searches all urls for a GitHub page. This is the default.
      -c, --sourcecode                 Open sourcecode gem
      -d, --doc                        Open documentation of gem
      -w, --webpage                    Open webpage of gem
      -t, --rubytoolbox                Open the rubytoolbox page for that gem
      -r, --rubygems                   Open the rubygems page for that gem


Let me hear some feedback about this.


**Platforms manually tested:** 

- [ ] aix
- [ ] cygwin
- [x] darwin
- [ ] macruby
- [ ] freebsd
- [ ] hpux
- [ ] java
- [ ] dalvik
- [ ] dotnet
- [x] linux
- [ ] mingw32
- [ ] netbsdelf
- [ ] openbsd
- [ ] bitrig
- [ ] solaris

# Tasks:

- [x] Describe the problem / feature
- [x] Write tests
- [x] Write code to solve the problem
- [x] Get code review from coworkers / friends

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

MatheusRich committed:
Add command web

MatheusRich committed:
Remove gem-web dependencies

MatheusRich committed:
Fix command web tests

MatheusRich committed:
Fix linter errors

MatheusRich committed:
Remove launchy

MatheusRich committed:
Remove fixtures

MatheusRich committed:
Update Manifest.txt

zoras commented on lib/rubygems/executor.rb:
```diff
@@ -0,0 +1,73 @@
+
+      # TODO: Make this cross-platform
+      def open_default_browser(uri)
+        system("xdg-open", uri)
```
Let's not hard code the executable opener command as it usually differs per OS

zoras reviewed (commented):
Nice 👍 
Do make sure to add opener command for all supported OS. 
Platform is already detected/known from here: https://github.com/rubygems/rubygems/blob/master/lib/rubygems/platform.rb#L79

MatheusRich committed:
Make open browser command cross-platform

zoras commented on lib/rubygems/executor.rb:
```diff
@@ -0,0 +1,86 @@
+  module Web
+    class Executor
+
+      OPEN_BROWSER_CMDS = {
```
👍

MatheusRich committed:
Adds open browser commands

MatheusRich committed:
Change open browser command hash to method

MatheusRich committed:
Updates open default browser commands

MatheusRich committed:
Add test to web command when platform is unsupported

MatheusRich committed:
Add tests for all platforms

MatheusRich committed:
Update usage on web command

MatheusRich committed:
Fix rubocop issues

zoras reviewed (commented):
Looks good Matheus. It would be great if we could do QA test on the various supported platforms as well as unsupported ones. Maybe you can try it out for various OS with VirtualBox?

MatheusRich committed:
Update web command behavior

Now this command uses the BROWSER env variable to open links. If it is unset, the command just outputs the URL

MatheusRich committed:
Fix web command tests

MatheusRich committed:
Remove github and rubytoolbox options from web command

MatheusRich committed:
Remove API call from web command

MatheusRich committed:
Support only some info from metadata

MatheusRich committed:
Refactor web command executor

MatheusRich committed:
Add API fallback to web command

MatheusRich committed:
Remove attr acessor from web command

MatheusRich committed:
Remove some useless code

MatheusRich committed:
Add minor refactoring
